File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/Model/IDOLanguageLibraryModel.swift
//
//  IDOLanguageLibraryModel.swift
//  protocol_channel
//
//  Created by hc on 2023/10/19.
//

import Foundation

// MARK: - IDOLanguageLibraryModel
@objcMembers
public class IDOLanguageLibraryModel: NSObject, IDOBaseModel {
    private var version: Int
    /// Currently used language
    public var useLang: Int
    /// Default language
    public var defaultLang: Int
    /// Number of fixed storage languages
    public var fixedLang: Int
    /// Maximum storage languages
    public var maxStorageLang: Int
    private var itemsLen: Int
    private var userLen: Int
    public var items: [IDOLanguageLibraryItem]?
    public var itemsUser: [IDOLanguageLibraryItem]?
    
    enum CodingKeys: String, CodingKey {
        case version = "version"
        case useLang = "use_lang"
        case defaultLang = "default_lang"
        case fixedLang = "fixed_lang"
        case maxStorageLang = "max_storage_lang"
        case itemsLen = "items_len"
        case userLen = "user_len"
        case items = "items"
        case itemsUser = "items_user"
    }
    
    public init(useLang: Int, defaultLang: Int, fixedLang: Int, maxStorageLang: Int, items: [IDOLanguageLibraryItem]?, itemsUser: [IDOLanguageLibraryItem]?) {
        self.version = 0
        self.useLang = useLang
        self.defaultLang = defaultLang
        self.fixedLang = fixedLang
        self.maxStorageLang = maxStorageLang
        self.itemsLen = items?.count ?? 0
        self.userLen = itemsUser?.count ?? 0
        self.items = items
        self.itemsUser = itemsUser
    }
    
    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}

// MARK: - IDOLanguageLibraryItem
@objcMembers
public class IDOLanguageLibraryItem: NSObject, Codable {
 
    /// Language type
    public var languageType: IDOLanguageType
    /// Language version number
    public var languageVersion: Int
    
    enum CodingKeys: String, CodingKey {
        case languageType = "language_type"
        case languageVersion = "language_version"
    }
    
    public init(languageType: IDOLanguageType, languageVersion: Int) {
        self.languageType = languageType
        self.languageVersion = languageVersion
    }
}


@objc
public enum IDOLanguageType: Int, Codable {
    case invalid = 0
    case chinese = 1
    case english = 2
    case french = 3
    case german = 4
    case italian = 5
    case spanish = 6
    case japanese = 7
    case polish = 8
    case czech = 9
    case romanian = 10
    case lithuanian = 11
    case dutch = 12
    case slovenian = 13
    case hungarian = 14
    case russian = 15
    case ukrainian = 16
    case slovak = 17
    case danish = 18
    case croatian = 19
    case indonesian = 20
    case korean = 21
    case hindi = 22
    case portuguese = 23
    case turkish = 24
    case thai = 25
    case vietnamese = 26
    case burmese = 27
    case filipino = 28
    case traditionalChinese = 29
    case greek = 30
    case arabic = 31
    case swedish = 32
    case finnish = 33
    case persian = 34
    case norwegian = 35
    case malay = 36
    case brazilianPortuguese = 37
    case bengali = 38
    case khmer = 39
    case serbian = 40
    case bulgaria = 41
    case hebrew = 42
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/model/IDOLanguageLibraryModel.kt
//
//  IDOLanguageLibraryModel.kt
//  protocol_channel
//
//  Created by hedongyang on 2023/10/30.
//

package com.idosmart.model

import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.lang.reflect.Type
import com.google.gson.annotations.JsonAdapter
import com.google.gson.annotations.SerializedName

///

// MARK: - IDOLanguageLibraryModel

open class IDOLanguageLibraryModel(
    useLang: Int,
    defaultLang: Int,
    fixedLang: Int,
    maxStorageLang: Int,
    items: List<IDOLanguageLibraryItem>?,
    itemsUser: List<IDOLanguageLibraryItem>?
) : IDOBaseModel {

    @SerializedName("version")
    private var version: Int = 0

    /// Currently used language
    @SerializedName("use_lang")
    var useLang: Int = useLang

    /// Default language
    @SerializedName("default_lang")
    var defaultLang: Int = defaultLang

    /// Number of fixed storage languages
    @SerializedName("fixed_lang")
    var fixedLang: Int = fixedLang

    /// Maximum storage languages
    @SerializedName("max_storage_lang")
    var maxStorageLang: Int = maxStorageLang
    @SerializedName("items_len")
    private var itemsLen: Int = items?.size ?: 0
    @SerializedName("user_len")
    private var userLen: Int = itemsUser?.size ?: 0

    @SerializedName("items")
    var items: List<IDOLanguageLibraryItem>? = items

    @SerializedName("items_user")
    var itemsUser: List<IDOLanguageLibraryItem>? = itemsUser




    override fun toJsonString(): String {
        return GsonBuilder().create().toJson(this).toString()
    }
}


// MARK: - IDOLanguageLibraryItem
class IDOLanguageLibraryItem(languageType: IDOLanguageType, languageVersion: Int) {
    /// Language type
    @SerializedName("language_type")
    @JsonAdapter(LanEnumToIntConverter::class)
    var languageType: IDOLanguageType = languageType

    /// Language version number
    @SerializedName("language_version")
    var languageVersion: Int = languageVersion

}

private class LanEnumToIntConverter : JsonSerializer<IDOLanguageType>,
    JsonDeserializer<IDOLanguageType> {

    override fun serialize(
        src: IDOLanguageType,
        typeOfSrc: Type,
        context: JsonSerializationContext
    ): JsonElement {
        return JsonPrimitive(src.raw)
    }

    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext
    ): IDOLanguageType {

        val value = json.asInt
        return IDOLanguageType.values().find { it.raw == value } ?: IDOLanguageType.invalid
    }
}

enum class IDOLanguageType(val raw: Int) {
    invalid(0),
    chinese(1),
    english(2),
    french(3),
    german(4),
    italian(5),
    spanish(6),
    japanese(7),
    polish(8),
    czech(9),
    romanian(10),
    lithuanian(11),
    dutch(12),
    slovenian(13),
    hungarian(14),
    russian(15),
    ukrainian(16),
    slovak(17),
    danish(18),
    croatian(19),
    indonesian(20),
    korean(21),
    hindi(22),
    portuguese(23),
    turkish(24),
    thai(25),
    vietnamese(26),
    burmese(27),
    filipino(28),
    traditionalChinese(29),
    greek(30),
    arabic(31),
    swedish(32),
    finnish(33),
    persian(34),
    norwegian(35),
    malay(36),
    brazilianPortuguese(37),
    bengali(38),
    khmer(39),
    serbian(40),
    bulgaria(41),
    hebrew(42);

    companion object {
        fun ofRaw(raw: Int): IDOLanguageType? {
            return IDOLanguageType.values().firstOrNull { it.raw == raw }
        }
    }
}
